pulsar/impl_partition_consumer.go

```diff
@@ -274,30 +309,25 @@ func (pc *partitionConsumer) ReceiveAsync(ctx context.Context, msgs chan<- Consu
 		case tmpMsg, ok := <-pc.subQueue:
 			if ok {
 				msgs <- tmpMsg
-				id := &pb.MessageIdData{}
-				err := proto.Unmarshal(tmpMsg.ID().Serialize(), id)
+
+				// track messages by message ID
+				err := pc.trackMessage(tmpMsg.ID())
 				if err != nil {
-					pc.log.WithError(err).Errorf("unserialize message id error:%s", err.Error())
 					return err
 				}
-				if pc.unAckTracker != nil {
-					pc.unAckTracker.Add(id)
-				}
+
+				// increase available permits
 				receivedSinceFlow++
-				if receivedSinceFlow >= highwater {
-					if err := pc.internalFlow(receivedSinceFlow); err != nil {
-						pc.log.Errorf("Send Flow cmd error:%s", err.Error())
-						return err
-					}
-					receivedSinceFlow = 0
+				err = pc.increaseAvailablePermits(receivedSinceFlow)
+				if err != nil {
+					return err
 				}
 				continue
 			}
 		case <-ctx.Done():
 			return ctx.Err()
 		}
 	}
```

Comment:
Please add a receiveAsync method, that returns only one message, and following the implementation of Java client.